File: koma-cli/src/main/java/org/koma/cli/command/RootCommand.java
package org.koma.cli.command;

import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.koma.cli.context.DI;
import org.koma.cli.provider.ManifestVersionProvider;
import picocli.CommandLine;

import java.util.concurrent.Callable;

@CommandLine.Command(
  name = "koma", mixinStandardHelpOptions = true,
  version = "koma 0.1",
  description = "Static Site generator",
  subcommands = {
    GenerateCommand.class,
    ServeCommand.class,
    WatchCommand.class,
    InitCommand.class
  },
  versionProvider = ManifestVersionProvider.class
)
@Slf4j
@Singleton
@RequiredArgsConstructor(onConstructor_ = {@Inject})
public class RootCommand implements Callable<Integer> {

  @Override
  public Integer call() {
    DI.INSTANCE.getBean(CommandLine.class).usage(System.out);
    return 0;
  }
}

File: koma-cli/src/main/kotlin/org/koma/cli/command/RootCommand.kt
package org.koma.cli.command

import com.github.ajalt.clikt.completion.completionOption
import com.github.ajalt.clikt.core.NoOpCliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.options.versionOption
import java.util.Properties
import org.koin.core.annotation.Single
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

@Single
class RootCommand : NoOpCliktCommand(), KoinComponent {
  private val generateCommand: GenerateCommand by inject()
  private val watchCommand: WatchCommand by inject()
  private val serverCommand: ServerCommand by inject()
  private val initCommand: InitCommand by inject()
  private val deployCommand: DeployCommand by inject()

  init {
    javaClass.getClassLoader().getResourceAsStream("META-INF/MANIFEST.MF")?.apply {
      val properties = Properties()
      properties.load(this)
      val version = properties.getValue("Implementation-Version")
      versionOption(version.toString())
    }
  }

  init {
    subcommands(generateCommand, initCommand, serverCommand, watchCommand, deployCommand)
  }

  init {
    completionOption()
  }

  override fun aliases(): Map<String, List<String>> =
      mapOf("g" to listOf("generate"), "i" to listOf("init"))
}
